src/Xamarin.Android.Build.Tasks/Utilities/MarshalMethodCecilAdapter.cs

```diff
@@ -0,0 +1,253 @@
+#nullable enable
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using Microsoft.Android.Build.Tasks;
+using Microsoft.Build.Utilities;
+using Mono.Cecil;
+using Xamarin.Android.Tools;
+
+namespace Xamarin.Android.Tasks;
+
+class MarshalMethodCecilAdapter
+{
+	public static NativeCodeGenStateCollection? GetNativeCodeGenStateCollection (TaskLoggingHelper log, ConcurrentDictionary<AndroidTargetArch, NativeCodeGenState>? nativeCodeGenStates)
+	{
+		if (nativeCodeGenStates is null) {
+			log.LogDebugMessage ($"No {nameof (NativeCodeGenState)} found");
+			return null;
+		}
+
+		var collection = new NativeCodeGenStateCollection ();
+
+		// Convert each architecture
+		foreach (var kvp in nativeCodeGenStates) {
+			var arch = kvp.Key;
+			var state = kvp.Value;
+			var obj = CreateNativeCodeGenState (arch, state);
+			collection.States.Add (arch, obj);
+		}
+
+		return collection;
+	}
+
+	static NativeCodeGenStateObject CreateNativeCodeGenState (AndroidTargetArch arch, NativeCodeGenState state)
+	{
+		var obj = new NativeCodeGenStateObject ();
+
+		if (state.Classifier is null)
+			return obj;
+
+		foreach (var group in state.Classifier.MarshalMethods) {
+			var methods = new List<MarshalMethodEntryObject> (state.Classifer.MarshalMethods.Count);
+
+			foreach (var method in group.Value) {
+				var entry = CreateEntry (method, state.ManagedMarshalMethodsLookupInfo);
+				methods.Add (entry);
+			}
+
+			obj.MarshalMethods.Add (group.Key, methods);
+		}
+
+		return obj;
+	}
+
+	static MarshalMethodEntryObject CreateEntry (MarshalMethodEntry entry, ManagedMarshalMethodsLookupInfo? info)
+	{
+		var obj = new MarshalMethodEntryObject (
+			declaringType: CreateDeclaringType (entry.DeclaringType),
+			implementedMethod: CreateMethod (entry.ImplementedMethod),
+			isSpecial: entry.IsSpecial,
+			jniTypeName: entry.JniTypeName,
+			jniMethodName: entry.JniMethodName,
+			jniMethodSignature: entry.JniMethodSignature,
+			nativeCallback: CreateMethod (entry.NativeCallback),
+			registeredMethod: CreateMethodBase (entry.RegisteredMethod)
+		);
+
+		if (info is not null) {
+			(uint assemblyIndex, uint classIndex, uint methodIndex) = info.GetIndex (entry.NativeCallback);
+
+			obj.NativeCallback.AssemblyIndex = assemblyIndex;
+			obj.NativeCallback.ClassIndex = classIndex;
+			obj.NativeCallback.MethodIndex = methodIndex;
+		}
+
+		return obj;
+	}
+
+	static MarshalMethodEntryTypeObject CreateDeclaringType (TypeDefinition type)
+	{
+		var cecilModule = type.Module;
+		var cecilAssembly = cecilModule.Assembly;
```

Comment:
Isn't the goal here to not use Cecil types? Do we need the full assembly type for MarshalMethodEntryModuleObject?